src/serializer/functions.js

```diff
@@ -125,49 +125,38 @@ export class Functions {
     return recordedAdditionalFunctions;
   }
 
-  // This will also handle postprocessing for PossiblyNormalCompletion
   _createAdditionalEffects(
     effects: Effects,
     fatalOnAbrupt: boolean,
     name: string,
     environmentRecordIdAfterGlobalCode: number,
     parentAdditionalFunction: FunctionValue | void = undefined
   ): AdditionalFunctionEffects | null {
+    let realm = this.realm;
     let [pncResult] = effects;
-    if (pncResult instanceof PossiblyNormalCompletion) {
-      // This is a join point for all the forked paths in pncResult
-      effects = Join.joinEffectsAndPromoteNestedReturnCompletions(
-        this.realm,
-        pncResult,
-        construct_empty_effects(this.realm)
+    if (pncResult instanceof PossiblyNormalCompletion || pncResult instanceof JoinedAbruptCompletions) {
+      // The completion is not the end of function execution, but a fork point for separate threads of control.
+      // The effects of all of these threads need to get joined up and rolled into the top level effects,
+      // so that applying the effects before serializing the body will fully initialize all variables and objects.
+      effects = realm.evaluateForEffects(
```

Comment:
Instead of evaluating and applying it, can't this be expressed as a pure join of the effects?

Reply:
I spent a lot of time looking at that option and failing. Join is not as self contained as you may think. In the join that results from "ob = { foo: 1, bar: 2 }; if (c) { ob.foo = 10; } else { ob. bar = 20; }; ob.foo;" then the joined value of ob.foo after the if-else is not c ? 10 : undefined, as you might conclude if you did the join in isolation, but rather c ? 10 : 1.

So, before we can do the Join.joinNestedEffects, we need apply the prior effects so that the join can find them in the objects.